[Bug]: opentelemetry-datadog does not compile with otel 0.29

### What happened?

After getting a renovate update to opentelemetry 0.29, we realized the opentelemetry-datadog crate did not get an update to work with the new version.

### API Version

N/A

### SDK Version

0.29

### What component are you working with?

_No response_

### Relevant log output

```shell
error[E0599]: the method `with_batch_config` exists for struct `BatchSpanProcessorBuilder<DatadogExporter>`, but its trait bounds were not satisfied
Error:    --> api/src/runtime/std/tracing.rs:112:14
    |
111 |           let batch_span_processor = BatchSpanProcessor::builder(exporter)
    |  ____________________________________-
112 | |             .with_batch_config(
    | |             -^^^^^^^^^^^^^^^^^ method cannot be called on `BatchSpanProcessorBuilder<DatadogExporter>` due to unsatisfied trait bounds
    | |_____________|
    |
    |
```

--- a/opentelemetry-datadog/CHANGELOG.md
+++ b/opentelemetry-datadog/CHANGELOG.md
@@ -2,6 +2,10 @@
 
 ## vNext
 
+- `DatadogExporter.export()` doesn't require mutablity anymore
+- Bump opentelemetry and opentelemetry_sdk versions to 0.29
+- Bump opentelemetry-http and opentelemetry-semantic-conventions versions to 0.29
+
 ## v0.16.0
 
 - Bump mvrv to 1.75.0

--- a/opentelemetry-datadog/Cargo.toml
+++ b/opentelemetry-datadog/Cargo.toml
@@ -36,7 +36,6 @@ reqwest = { version = "0.12", default-features = false, optional = true }
 surf = { version = "2.0", default-features = false, optional = true }
 thiserror = "2.0"
 http = "1"
-futures-core = "0.3"
 ryu = "1"
 itoa = "1"
 ahash = { version = "0.8", optional = true }

--- a/opentelemetry-datadog/Cargo.toml
+++ b/opentelemetry-datadog/Cargo.toml
@@ -26,10 +26,10 @@ intern-std = []
 
 [dependencies]
 indexmap = "2.0"
-opentelemetry = { version = "0.28" }
-opentelemetry_sdk = { version = "0.28", features = ["trace"] }
-opentelemetry-http = { version = "0.28" }
-opentelemetry-semantic-conventions = { version = "0.28" }
+opentelemetry = { workspace = true }
+opentelemetry_sdk = { workspace = true, features = ["trace"] }
+opentelemetry-http = { workspace = true }
+opentelemetry-semantic-conventions = { workspace = true }
 rmp = "0.8"
 url = "2.2"
 reqwest = { version = "0.12", default-features = false, optional = true }
@@ -46,7 +46,7 @@ async-trait = "0.1"
 base64 = "0.22"
 bytes = "1"
 futures-util = { version = "0.3", default-features = false, features = ["io"] }
-opentelemetry_sdk = { version = "0.28", features = ["trace", "testing"] }
+opentelemetry_sdk = { workspace = true, features = ["trace", "testing"] }
 criterion = "0.5"
 rand = "0.9"
 hyper = "1"

--- a/opentelemetry-datadog/benches/datadog_exporter.rs
+++ b/opentelemetry-datadog/benches/datadog_exporter.rs
@@ -196,7 +196,7 @@ fn generate_traces(number_of_traces: usize, spans_per_trace: usize) -> Vec<SpanD
 }
 
 fn criterion_benchmark(c: &mut Criterion) {
-    let mut exporter = new_pipeline()
+    let exporter = new_pipeline()
         .with_service_name("trace-demo")
         .with_api_version(ApiVersion::Version05)
         .with_http_client(DummyClient)

--- a/opentelemetry-datadog/examples/agent_sampling.rs
+++ b/opentelemetry-datadog/examples/agent_sampling.rs
@@ -57,15 +57,14 @@ impl ShouldSample for AgentBasedSampler {
 }
 
 fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
-    #[allow(deprecated)]
+    let mut config = trace::Config::default();
+    config.sampler = Box::new(AgentBasedSampler);
+    config.id_generator = Box::new(RandomIdGenerator::default());
+
     let provider = new_pipeline()
         .with_service_name("agent-sampling-demo")
         .with_api_version(ApiVersion::Version05)
-        .with_trace_config(
-            trace::Config::default()
-                .with_sampler(AgentBasedSampler)
-                .with_id_generator(RandomIdGenerator::default()),
-        )
+        .with_trace_config(config)
         .install_simple()?;
     global::set_tracer_provider(provider.clone());
     let scope = InstrumentationScope::builder("opentelemetry-datadog-demo")

--- a/opentelemetry-datadog/src/exporter/mod.rs
+++ b/opentelemetry-datadog/src/exporter/mod.rs
@@ -6,15 +6,13 @@ pub use model::Error;
 pub use model::FieldMappingFn;
 
 use crate::exporter::model::FieldMapping;
-use futures_core::future::BoxFuture;
 use http::{Method, Request, Uri};
-use opentelemetry::Key;
-use opentelemetry::{trace::TraceError, KeyValue};
+use opentelemetry::{Key, KeyValue};
 use opentelemetry_http::{HttpClient, ResponseExt};
 use opentelemetry_sdk::{
     error::{OTelSdkError, OTelSdkResult},
     resource::{ResourceDetector, SdkProvidedResourceDetector},
-    trace::{Config, SdkTracerProvider},
+    trace::{Config, SdkTracerProvider, TraceError},
     trace::{SpanData, SpanExporter},
     Resource,
 };
@@ -424,14 +422,14 @@ async fn send_request(
 
 impl SpanExporter for DatadogExporter {
     /// Export spans to datadog-agent
-    fn export(&mut self, batch: Vec<SpanData>) -> BoxFuture<'static, OTelSdkResult> {
+    async fn export(&self, batch: Vec<SpanData>) -> OTelSdkResult {
         let request = match self.build_request(batch) {
             Ok(req) => req,
-            Err(err) => return Box::pin(std::future::ready(Err(err))),
+            Err(err) => return Err(err),
         };
 
         let client = self.client.clone();
-        Box::pin(send_request(client, request))
+        send_request(client, request).await
     }
     fn set_resource(&mut self, resource: &Resource) {
         self.resource = Some(resource.clone());

--- a/opentelemetry-datadog/src/exporter/model/mod.rs
+++ b/opentelemetry-datadog/src/exporter/model/mod.rs
@@ -1,9 +1,8 @@
 use crate::exporter::ModelConfig;
 use http::uri;
-use opentelemetry::trace::ExportError;
 use opentelemetry_sdk::{
     trace::{self, SpanData},
-    Resource,
+    ExportError, Resource,
 };
 use std::fmt::Debug;
 use url::ParseError;
@@ -45,7 +44,7 @@ static DD_MEASURED_KEY: &str = "_dd.measured";
 /// use opentelemetry::global;
 /// use opentelemetry_datadog::{ApiVersion, new_pipeline};
 ///
-/// fn main() -> Result<(), opentelemetry::trace::TraceError> {
+/// fn main() -> Result<(), opentelemetry_sdk::trace::TraceError> {
 ///     let provider = new_pipeline()
 ///         .with_service_name("my_app")
 ///         .with_api_version(ApiVersion::Version05)

--- a/opentelemetry-datadog/src/lib.rs
+++ b/opentelemetry-datadog/src/lib.rs
@@ -48,7 +48,7 @@
 //! ```
 //!
 //! ```no_run
-//! # fn main() -> Result<(), opentelemetry::trace::TraceError> {
+//! # fn main() -> Result<(), opentelemetry_sdk::trace::TraceError> {
 //! let provider = opentelemetry_datadog::new_pipeline()
 //!     .install_batch()?;
 //! # Ok(())
@@ -122,16 +122,15 @@
 //!     }
 //! }
 //!
-//!     #[allow(deprecated)]
+//!     let mut config = trace::Config::default();
+//!     config.sampler = Box::new(Sampler::AlwaysOn);
+//!     config.id_generator = Box::new(RandomIdGenerator::default());
+//!
 //!     let provider = new_pipeline()
 //!         .with_service_name("my_app")
 //!         .with_api_version(ApiVersion::Version05)
 //!         .with_agent_endpoint("http://localhost:8126")
-//!         .with_trace_config(
-//!             trace::config()
-//!                 .with_sampler(Sampler::AlwaysOn)
-//!                 .with_id_generator(RandomIdGenerator::default())
-//!         )
+//!         .with_trace_config(config)
 //!         .install_batch().unwrap();
 //!     global::set_tracer_provider(provider.clone());
 //!

